File: codewars/javascript/validBraces.js
function validBraces(braces) {
  if (braces.length === 0) return true;

  let parenthesisStack = [];

  for (let i = 0; i < braces.length; i++) {
    if (
      (braces[i] === ")" || braces[i] === "]" || braces[i] === "}") &&
      parenthesisStack.length === 0
    ) {
      return false;
    } else if (
      (braces[i] === ")" &&
        parenthesisStack[parenthesisStack.length - 1] === "(") ||
      (braces[i] === "]" &&
        parenthesisStack[parenthesisStack.length - 1] === "[") ||
      (braces[i] === "}" &&
        parenthesisStack[parenthesisStack.length - 1] === "{")
    ) {
      parenthesisStack.pop();
    } else {
      parenthesisStack.push(braces[i]);
    }
  }
  return parenthesisStack.length === 0;
}

File: codewars/dart/validBraces.dart
bool validBraces(String braces) {
  if (braces.isEmpty) return true;

  List<String> parenthesisStack = [];

  for (int i = 0; i < braces.length; i++) {
    if ((braces[i] == ')' || braces[i] == ']' || braces[i] == '}') &&
        parenthesisStack.isEmpty) {
      return false;
    } else if ((braces[i] == ')' && parenthesisStack.last == '(') ||
        (braces[i] == ']' && parenthesisStack.last == '[') ||
        (braces[i] == '}' && parenthesisStack.last == '{')) {
      parenthesisStack.removeLast();
    } else {
      parenthesisStack.add(braces[i]);
    }
  }
  return parenthesisStack.isEmpty;
}
